File: frontend/src/api/chat-integrations.ts
// Added: Chat integration API client for team chat webhook management (TMAIL-129)

import { apiClient } from './client';

// Added: Chat platform types matching backend enum
export type ChatPlatform = 'slack' | 'teams' | 'google_chat' | 'discord' | 'custom';

export interface ChatIntegration {
  id: string;
  user_id: string;
  platform: ChatPlatform;
  webhook_url: string;
  channel_name: string | null;
  notify_on_receive: boolean;
  notify_on_send: boolean;
  notify_on_mention: boolean;
  filter_from: string | null;
  filter_subject: string | null;
  active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateChatIntegrationRequest {
  platform: ChatPlatform;
  webhook_url: string;
  channel_name?: string;
  notify_on_receive?: boolean;
  notify_on_send?: boolean;
  notify_on_mention?: boolean;
  filter_from?: string;
  filter_subject?: string;
}

export interface UpdateChatIntegrationRequest {
  webhook_url?: string;
  channel_name?: string;
  notify_on_receive?: boolean;
  notify_on_send?: boolean;
  notify_on_mention?: boolean;
  filter_from?: string;
  filter_subject?: string;
  active?: boolean;
}

export interface TestResult {
  success: boolean;
  message: string;
}

// PURPOSE: List all chat integrations for the current user
export async function listChatIntegrations(): Promise<ChatIntegration[]> {
  return apiClient.get<ChatIntegration[]>('/chat-integrations');
}

// PURPOSE: Create a new chat integration
export async function createChatIntegration(data: CreateChatIntegrationRequest): Promise<ChatIntegration> {
  return apiClient.post<ChatIntegration>('/chat-integrations', data);
}

// PURPOSE: Get a single chat integration by ID
export async function getChatIntegration(id: string): Promise<ChatIntegration> {
  return apiClient.get<ChatIntegration>(`/chat-integrations/${id}`);
}

// PURPOSE: Update an existing chat integration
export async function updateChatIntegration(id: string, data: UpdateChatIntegrationRequest): Promise<ChatIntegration> {
  return apiClient.put<ChatIntegration>(`/chat-integrations/${id}`, data);
}

// PURPOSE: Delete a chat integration
export async function deleteChatIntegration(id: string): Promise<void> {
  await apiClient.delete(`/chat-integrations/${id}`);
}

// PURPOSE: Send a test notification to the chat integration's webhook
export async function testChatIntegration(id: string): Promise<TestResult> {
  return apiClient.post<TestResult>(`/chat-integrations/${id}/test`, {});
}

File: backend/src/handlers/chat_integrations.rs
// Added: Chat integration management handlers for TMAIL-129
// PURPOSE: CRUD endpoints for managing team chat webhook integrations
// EXTERNAL: Uses chat_notifier service for test message delivery

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};

use crate::error::AppError;
use crate::models::chat_integration::{
    ChatIntegration, CreateChatIntegrationRequest, UpdateChatIntegrationRequest,
};
use crate::services::auth_service::Claims;
use crate::services::chat_notifier;
use crate::state::AppState;

/// PURPOSE: List all chat integrations for the authenticated user
/// GET /api/chat-integrations
pub async fn list_chat_integrations(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
) -> Result<Json<Vec<ChatIntegration>>, AppError> {
    let user_id = parse_user_id(&claims)?;
    let integrations = ChatIntegration::find_by_user(&state.db, user_id).await?;
    Ok(Json(integrations))
}

/// PURPOSE: Create a new chat integration
/// POST /api/chat-integrations
/// CONSTRAINTS: webhook_url must be a valid HTTP(S) URL
pub async fn create_chat_integration(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
    Json(body): Json<CreateChatIntegrationRequest>,
) -> Result<(StatusCode, Json<ChatIntegration>), AppError> {
    let user_id = parse_user_id(&claims)?;

    // Added: Validate URL format
    if !body.webhook_url.starts_with("https://") && !body.webhook_url.starts_with("http://") {
        return Err(AppError::BadRequest(
            "Webhook URL must start with http:// or https://".to_string(),
        ));
    }

    let integration = ChatIntegration::create(&state.db, user_id, &body).await?;
    Ok((StatusCode::CREATED, Json(integration)))
}

/// PURPOSE: Get a single chat integration by ID
/// GET /api/chat-integrations/:id
pub async fn get_chat_integration(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
    Path(id): Path<uuid::Uuid>,
) -> Result<Json<ChatIntegration>, AppError> {
    let user_id = parse_user_id(&claims)?;
    let integration = ChatIntegration::find_by_id(&state.db, id, user_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Chat integration not found".to_string()))?;
    Ok(Json(integration))
}

/// PURPOSE: Update an existing chat integration's configuration
/// PUT /api/chat-integrations/:id
pub async fn update_chat_integration(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
    Path(id): Path<uuid::Uuid>,
    Json(body): Json<UpdateChatIntegrationRequest>,
) -> Result<Json<ChatIntegration>, AppError> {
    let user_id = parse_user_id(&claims)?;

    // Added: Validate URL if provided
    if let Some(ref url) = body.webhook_url {
        if !url.starts_with("https://") && !url.starts_with("http://") {
            return Err(AppError::BadRequest(
                "Webhook URL must start with http:// or https://".to_string(),
            ));
        }
    }

    let integration = ChatIntegration::update(&state.db, id, user_id, &body)
        .await?
        .ok_or_else(|| AppError::NotFound("Chat integration not found".to_string()))?;
    Ok(Json(integration))
}

/// PURPOSE: Delete a chat integration
/// DELETE /api/chat-integrations/:id
pub async fn delete_chat_integration(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
    Path(id): Path<uuid::Uuid>,
) -> Result<StatusCode, AppError> {
    let user_id = parse_user_id(&claims)?;
    let deleted = ChatIntegration::delete(&state.db, id, user_id).await?;
    if deleted {
        Ok(StatusCode::NO_CONTENT)
    } else {
        Err(AppError::NotFound("Chat integration not found".to_string()))
    }
}

/// PURPOSE: Send a test notification to the chat integration's webhook
/// POST /api/chat-integrations/:id/test
/// NOTE: Uses the chat_notifier service to format and send a test message
pub async fn test_chat_integration(
    State(state): State<AppState>,
    axum::Extension(claims): axum::Extension<Claims>,
    Path(id): Path<uuid::Uuid>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = parse_user_id(&claims)?;
    let integration = ChatIntegration::find_by_id(&state.db, id, user_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Chat integration not found".to_string()))?;

    // Added: Send test notification via the chat notifier service
    let result = chat_notifier::send_test_notification(
        &integration.platform,
        &integration.webhook_url,
    )
    .await;

    match result {
        Ok(_) => Ok(Json(serde_json::json!({
            "success": true,
            "message": "Test notification sent successfully"
        }))),
        Err(error_message) => Ok(Json(serde_json::json!({
            "success": false,
            "message": error_message
        }))),
    }
}

fn parse_user_id(claims: &Claims) -> Result<uuid::Uuid, AppError> {
    claims
        .sub
        .parse()
        .map_err(|_| AppError::Internal(anyhow::anyhow!("Invalid user ID in JWT claims")))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::services::auth_service::Claims;

    #[test]
    fn test_parse_user_id_valid() {
        let claims = Claims {
            sub: uuid::Uuid::new_v4().to_string(),
            username: "test@example.com".into(),
            is_admin: false,
            is_compliance_officer: false,
            exp: 0,
            iat: 0,
        };
        assert!(parse_user_id(&claims).is_ok());
    }

    #[test]
    fn test_parse_user_id_invalid() {
        let claims = Claims {
            sub: "not-a-uuid".into(),
            username: "test@example.com".into(),
            is_admin: false,
            is_compliance_officer: false,
            exp: 0,
            iat: 0,
        };
        assert!(parse_user_id(&claims).is_err());
    }
}
